python/ray/serve/backend_state_manager.py

```diff
@@ -980,15 +1103,26 @@ def _completed_goals(self) -> List[GoalId]:
             del self._target_replicas[backend_tag]
             del self._target_versions[backend_tag]
 
-        return [goal for goal in completed_goals if goal]
+        return [goal for goal in completed_goals
+                if goal], [goal for goal in failed_to_start_goals if goal]
 
     def update(self) -> bool:
         """Updates the state of all running replicas to match the goal state.
         """
         self._scale_all_backends()
 
-        for goal_id in self._completed_goals():
+        (complete_goal_ids,
+         failed_to_start_goal_ids) = self._check_completed_goals()
+        for goal_id in complete_goal_ids:
             self._goal_manager.complete_goal(goal_id)
+        for goal_id in failed_to_start_goal_ids:
+            self._goal_manager.complete_goal(
+                goal_id,
+                # TODO:(jiaodong) Do we want to surface original constructor
+                # exception from other actor instead ? Might need controller
+                # level global exception handler
+                RuntimeError(f"Deployment with goal_id {goal_id} "
+                             f"failed to start. See worker exception above."))
```

Comment:
This shouldn't be too hard, should it? Just log it in one of the per-backend datatstructure (`most_recent_constructor_failure`)

Reply:
yep it shouldn't be hard, as diff summary said i was also thinking of setting it at event loop level for things like replica_tag: exception. We already had it at worker.py, just need to do one more propagation.